### PR TITLE
1-Docker

### PR DESCRIPTION
**Pull request by dahachm**
@konstantin009 @Shlepanets

**Docker containers with hadoop**

https://hub.docker.com/r/dahachm/hadoop

**Headnode** (namenode):
Get image:
```
$ docker pull dahachm/hadoop:headnode
```
Start container:
```
$ docker run --name headnode -dP -p 8088:8088 dahachm/hadoop:headnode
```

Access ResourceManager WebUI at *port 8080* on your host machine.

**Worker** (slave node):
Get container:
```
$ docker pull dahachm/hadoop:headnode
```
Start container:
```
$ docker run --name worker -dP -p 8042:8042 dahachm/hadoop:worker
```

Access NodeManager WebUI at *port 8042* on your host machine.

**Docker-compose file**

**docker-compose.yml** - for building up these containers and start it properly.

 Start composing:
```
$ docker-compose -f docker-compose.yml up -d 
```

### CODE REVIEW

File: 1-Docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+version: '3'
+
+services:
+  headnode:
+    image: dahachm/hadoop:headnode
```
Comment (Shlepanets) [inline 1-Docker/docker-compose.yml]: Tags should be used to track versions of one image (ex. [Jenkins DockerHub](https://hub.docker.com/_/jenkins?tab=tags&page=1&ordering=last_updated) ), not to distinguish between headnode and worker.
There should be two separate images, i.e.: dahachm/hadoop_headnode:1.2 (or latest) and ahachm/hadoop_worker:1.15 (or latest)

File: 1-Docker/headnode_dockerfile [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+ENV JAVA_HOME /usr/lib/jvm/java-1.8-openjdk/jre
+ENV HADOOP_HOME /usr/local/hadoop-3.1.2
+
+COPY scripts/set-hadoop-conf.sh /usr/local/set-hadoop-conf.sh
+
+RUN /usr/local/set-hadoop-conf.sh
+
+RUN chown root:hadoop $HADOOP_HOME && \
+    chown root:hadoop $HADOOP_HOME/* && \
```
Comment (Shlepanets) [inline 1-Docker/headnode_dockerfile:28]: what is the expected result after that command?

File: 1-Docker/docker-compose.yml (hunk 1 above)
Comment (dahachm) [reply 1-Docker/docker-compose.yml]: Okay, was guessing is it right to put them in the same repo with different tags

The reason why I put them together was about the fact that thay are in the same task and they both are hadoop containers

Should I move one to another repo?

File: 1-Docker/scripts/set-hosts-headnode.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (Shlepanets) [inline 1-Docker/scripts/set-hosts-headnode.sh]: I really like that you've used awk here, Konstantin Radostev will be very proud of you =)
But jokes aside, what if, let's say, you would need to launch more than one worker in a cluster? Or if I decide to launch worker at first, and then headnode?
Is there a better way to communicate between containers?
Hint: you can already ping from one container to another, even without executing those set-hosts scripts, and without changing /etc/hosts. How?

**Review by Shlepanets — COMMENTED** [review]
General comments: 
- there is no need to expose that amount of ports. Usually, you should expose only necessary ports which will be accessed from "outside" network, such as Resource Manager UI (8088), NodeManager UI (8042), NameNode UI
- you should use "git squash" or similar command, in order to avoid merging 10 commits with the same "Update README.md"

File: 1-Docker/headnode_dockerfile (hunk 2 above)
Comment (dahachm) [reply 1-Docker/headnode_dockerfile:28]: */usr/local/set-hadoop-conf.sh* - this script 
downloads hadoop config files (hadoop-env.sh, core-site.xml, hdfs-site.xml, yarn-site.xml) - the same that we had on Linux Exam - and set required parameters (JAVA_HOME, HADOOP_HOME, hostnames and dirs)

As we untared hadoop archive under root, *I needed to change the owning group of the hadoop working files and directories (to escape Permission denied fault)*

So in these lines I change the owning user and group of $HADOOP_HOME directory recursively (when I tested these commands, somehow **-r** parameter didn't work, so I put it in two lines)

File: 1-Docker/scripts/set-hosts-headnode.sh (hunk 3 above)
Comment (dahachm) [reply 1-Docker/scripts/set-hosts-headnode.sh]: Yes, you're right! There is much easier way to find out IP address of other nodes!

From docker documentation:
> By default Compose sets up a single network for your app. Each container for a service joins the default network and is both reachable by other containers on that network, and discoverable by them at a hostname identical to the container name.

So I can get IP address of other node by using **dig** (from bind-utils), for instance

Here I tried to resolve **headnode** name from **worker**, it works:
Screenshot_5

So there is no need to add smth to /etc/hosts

**Comment by dahachm** [comment]
@Shlepanets , thank you a lot!

I will consider all your comments and update ASAP!

File: 1-Docker/docker-compose.yml (hunk 1 above)
Comment (Shlepanets) [reply 1-Docker/docker-compose.yml]: You can do it, if you'll have spare time, but this is not necessary to pass hometask =)

File: 1-Docker/headnode_dockerfile (hunk 2 above)
Comment (Shlepanets) [reply 1-Docker/headnode_dockerfile:28]: 1. Why didn't "chown -r" work?
2. Does this command changes owner for **all** files and directories inside $HADOOP_HOME/? 
chown root:hadoop $HADOOP_HOME/*

**Comment by dahachm** [comment]
@Shlepanets,

1. Removed *set-hosts* scripts and updated *start-*.sh* scripts (as there is no need to edit /etc/hosts)

2. Updated Docker Hub repo, now there are 2 images in different repos (https://hub.docker.com/repositories):
   **headnode**:  `docker pull dahachm/hadoop-headnode:1.0`
   **worker**:  `docker pull dahachm/hadoop-worker:1.0`
 
3. Updated *docker-compose.yml*:  deleted excess posrts

4. Rebased history of comits. I do remember, that it's a bad idea to do so after pushing in remote repository, but I will keep in mind  that all the rebase stuff must be done before pushing :)

File: 1-Docker/headnode_dockerfile (hunk 2 above)
Comment (dahachm) [reply 1-Docker/headnode_dockerfile:28]: > Why didn't "chown -r" work?

*/bin/sh* doesn't support **-r** parameter. I just checked its "help page", it does support **-R** parameter :'D 
So I really had a chance to make it one-line comand 

> Does this command changes owner for all files and directories inside $HADOOP_HOME/?
chown root:hadoop $HADOOP_HOME/*

No, it apply changes only for **1-st layer** of $HADOOP_HOME dir. Using **-R** would do the job, but there was no need for that :)

Hadoop was complaing on having lack of permissions for writing in home dir. Now I'm thinking that change owner of $HADOOP_HOME dir and $HADOOP_HOME/logs (after fomatting of FS) later  would be just enough

**Pull request merged**